src/core/symbology/qgssvgcache.cpp

```diff
@@ -406,77 +409,75 @@ QByteArray QgsSvgCache::getImageData( const QString &path ) const
     return mMissingSvg;
   }
 
-  // the url points to a remote resource, download it!
-  QNetworkReply *reply = nullptr;
+  QMutexLocker locker( &mMutex );
+
+  // already a request in progress for this url
+  if ( mPendingRemoteUrls.contains( path ) )
+    return mFetchingSvg;
 
-  // The following code blocks until the file is downloaded...
-  // TODO: use signals to get reply finished notification, in this moment
-  // it's executed while rendering.
-  while ( true )
+  if ( mRemoteContentCache.contains( path ) )
   {
-    QgsDebugMsg( QString( "get svg: %1" ).arg( svgUrl.toString() ) );
-    QNetworkRequest request( svgUrl );
-    request.setAttribute( QNetworkRequest::CacheLoadControlAttribute, QNetworkRequest::PreferCache );
-    request.setAttribute( QNetworkRequest::CacheSaveControlAttribute, true );
+    // already fetched this content - phew. Just return what we already got.
+    return *mRemoteContentCache[ path ];
+  }
 
-    reply = QgsNetworkAccessManager::instance()->get( request );
-    connect( reply, &QNetworkReply::downloadProgress, this, &QgsSvgCache::downloadProgress );
+  mPendingRemoteUrls.insert( path );
+  //fire up task to fetch image in background
+  QNetworkRequest request( svgUrl );
+  request.setAttribute( QNetworkRequest::CacheLoadControlAttribute, QNetworkRequest::PreferCache );
+  request.setAttribute( QNetworkRequest::CacheSaveControlAttribute, true );
 
-    //emit statusChanged( tr( "Downloading svg." ) );
+  QgsNetworkContentFetcherTask *task = new QgsNetworkContentFetcherTask( request );
+  connect( task, &QgsNetworkContentFetcherTask::fetched, this, [this, task, path]
+  {
+    QMutexLocker locker( &mMutex );
 
-    // wait until the image download finished
-    // TODO: connect to the reply->finished() signal
-    while ( !reply->isFinished() )
+    QNetworkReply *reply = task->reply();
+    if ( !reply )
     {
-      QCoreApplication::processEvents( QEventLoop::ExcludeUserInputEvents, 500 );
+      // canceled
+      QMetaObject::invokeMethod( const_cast< QgsSvgCache * >( this ), "onRemoteSvgFetched", Qt::QueuedConnection, Q_ARG( QString, path ), Q_ARG( bool, false ) );
```

Comment:
Should this method be un-consted instead?

Reply:
It's called from many const methods though, and i think they are validly const. It's a fine line with caches...